llvm/lib/Analysis/DXILResource.cpp

```diff
@@ -331,6 +336,249 @@ std::pair<uint32_t, uint32_t> ResourceInfo::getAnnotateProps() const {
+    if (HandleTy->getNumTypeParameters() != 1 ||
+        HandleTy->getNumIntParameters() != (IsTyped ? 3 : 2)) {
+      diagnoseHandle(CI, Twine("Invalid buffer target type"));
+      return nullptr;
+    }
+
+    Type *ElTy = HandleTy->getTypeParameter(0);
+    unsigned IsWriteable = HandleTy->getIntParameter(0);
+    unsigned IsROV = HandleTy->getIntParameter(1);
+    bool IsSigned = IsTyped && HandleTy->getIntParameter(2);
+
+    ResourceClass RC = IsWriteable ? ResourceClass::UAV : ResourceClass::SRV;
+    ResourceKind Kind;
+    if (IsTyped)
+      Kind = ResourceKind::TypedBuffer;
+    else if (ElTy->isIntegerTy(8))
+      Kind = ResourceKind::RawBuffer;
+    else
+      Kind = ResourceKind::StructuredBuffer;
+
+    // TODO: We need to lower to a typed pointer, can we smuggle the type
+    // through?
+    Value *Symbol = UndefValue::get(PointerType::getUnqual(Context));
+    // TODO: We don't actually keep track of the name right now...
+    StringRef Name = "";
+
+    auto [It, Success] = Resources.try_emplace(CI, RC, Kind, Symbol, Name);
+    assert(Success && "Mapping the same CallInst again?");
+    (void)Success;
+    // We grab a pointer into the map's storage, which isn't generally safe.
+    // Since we're just using this to fill in the info the map won't mutate and
+    // the pointer stays valid for as long as we need it to.
+    ResourceInfo *RI = &(It->second);
+
+    if (RI->isUAV())
+      // TODO: We need analysis for GloballyCoherent and HasCounter
+      RI->setUAV(false, false, IsROV);
+
+    if (RI->isTyped()) {
+      dxil::ElementType ET = toDXILElementType(ElTy, IsSigned);
+      uint32_t Count = 1;
+      if (auto *VTy = dyn_cast<FixedVectorType>(ElTy))
+        Count = VTy->getNumElements();
+      RI->setTyped(ET, Count);
+    } else if (RI->isStruct()) {
+      const DataLayout &DL = M.getDataLayout();
+
+      // This mimics what DXC does. Notably, we only ever set the alignment if
+      // the type is actually a struct type.
+      uint32_t Stride = DL.getTypeAllocSize(ElTy);
+      MaybeAlign Alignment;
+      if (auto *STy = dyn_cast<StructType>(ElTy))
+        Alignment = DL.getStructLayout(STy)->getAlignment();
+      RI->setStruct(Stride, Alignment);
+    }
+
+    return RI;
+  }
+
+  ResourceInfo *mapHandleIntrin(CallInst *CI) {
+    FunctionType *FTy = CI->getFunctionType();
+    Type *RetTy = FTy->getReturnType();
+    auto *HandleTy = dyn_cast<TargetExtType>(RetTy);
+    if (!HandleTy) {
+      diagnoseHandle(CI, "dx.handle.fromBinding requires target type");
+      return nullptr;
+    }
+
+    StringRef TypeName = HandleTy->getName();
+    if (TypeName == "dx.TypedBuffer") {
+      return mapBufferType(CI, HandleTy, /*IsTyped=*/true);
+    } else if (TypeName == "dx.RawBuffer") {
+      return mapBufferType(CI, HandleTy, /*IsTyped=*/false);
+    } else if (TypeName == "dx.CBuffer") {
+      // TODO: implement
+      diagnoseHandle(CI, "dx.CBuffer handles are not implemented yet");
```

Comment:
Could we use ptr with special address space for RawBuffer and CBuffer instead of handles for clang codeGen output?
Then dx.RawBuffer could be dx.StructuredBuffer, since it doesn't share between RawBuffer and StructuredBuffer anymore.

Reply:
We could potentially use that approach for RawBuffer, but that's something we'd need to design and it seems awkward to have a separate code path for RawBuffer vs the other buffer and texture types. I'm not sure if it would be worth it or not.

CBuffer probably shouldn't be done like that - it's much more akin to TypedBuffer in that it can only be accessed by 16 byte loads and it has all of the special layout stuff to deal with.

Reply:
For hlsl
```
struct A {
  float a;
};
ConstantBuffer<A> cb;
ByteAddressBuffer bab;
Buffer<float> tb;
```
dxc -spirv will generate

```
%type_ConstantBuffer_A                     = OpTypeStruct %float
%_ptr_Uniform_type_ConstantBuffer_A        = OpTypePointer Uniform %type_ConstantBuffer_A
 %cb                                       = OpVariable %_ptr_Uniform_type_ConstantBuffer_A Uniform

%_runtimearr_uint                          = OpTypeRuntimeArray %uint
%type_ByteAddressBuffer                    = OpTypeStruct %_runtimearr_uint
%_ptr_Uniform_type_ByteAddressBuffer       = OpTypePointer Uniform %type_ByteAddressBuffer
 %bab                                      = OpVariable %_ptr_Uniform_type_ByteAddressBuffer Uniform

%type_buffer_image                      = OpTypeImage %float Buffer 2 0 0 1 R32f
%_ptr_UniformConstant_type_buffer_image = OpTypePointer UniformConstant %type_buffer_image
 %tb                                    = OpVariable %_ptr_UniformConstant_type_buffer_image UniformConstant
```

Both RawBuffer and ConstantBuffer are pointers of struct while TypedBuffer is pointer of OpTypeImage.
So CBuffer is closer to RawBuffer instead of TypedBuffer.

Reply:
I don't know about SPIR-V, but just using a struct for a cbuffer in LLVM IR opens up a bunch of complexity about the layout. Consider:

```hlsl
cbuffer FunnyLayout { 
   float2 first;
   float arr[2];
   float2 last;
};
```

This is layed out like so in a cbuffer, in 3 rows (excuse the ascii art):
```
 0        4        8       12
 +--------+--------+--------+--------+
 |      first      |        |        |
 +--------+--------+--------+--------+
16       20       24       28
 +--------+--------+--------+--------+
 | arr[0] |        |        |        |
 +--------+--------+--------+--------+
32       36       40       44
 +--------+--------+--------+--------+
 | arr[1] |      last       |        |
 +--------+--------+--------+--------+
```

To access "last", we do a cbuffer load with index 2 and take the second element.

There are two problems with using a pointer to an llvm struct type for this. First, the layout of `{<2 x float>, [2 x float], <2 x float>}` is as follows:
```
0        4        8       12
 +--------+--------+--------+--------+
 |      first      | arr[0] | arr[1] |
 +--------+--------+--------+--------+
16       20
 +--------+--------+
 |      last       |
 +--------+--------+
```

Second, even if you were to construct a type with a bunch of padding so that it's layed out like a cbuffer, we'd have to enforce that all loads and stores are of exactly 16 bytes on a 16 byte boundary. This is trivial if we have an intrinsic for it, but requires extra validation and logic in various places and special knowledge in optimizations if we just use plain loads and stores.

Reply:
We can use https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/DirectX/CBufferDataLayout.h to calculate the  layout or size for cbuffer.
If we create the 16bytes intrinsic at clang codeGen, SPIRV will need more work to make it back to struct.
Or do you mean clang codeGen generate struct pointer, the 16bytes intrinsic just generated at DirectX backend?

Reply:
FWIW clang codegen for SPIR-V doesn't necessarily need to generate the same thing for SPIR-V and DirectX here. The path that lowers to target intrinsics has hooks per target - see [CGHLSLRuntime::convertHLSLSpecificType](https://github.com/llvm/llvm-project/pull/97362/files#diff-9550d9b19bd0f2011683d22a8de19dd126d5f36b5befcfa0a47b95652169f0cf) in #97362 and [TargetCodeGenInfo::getOpenCLType](https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/Targets/SPIR.cpp#L175) to see what I mean. If it ends up making more sense to do things differently for SPIR-V or other targets that aren't DXIL, we have the tools to deal with that.

Even so, there are still issues with "just" making cbuffer a pointer to a struct in a specific address space. We have to do analysis to figure out the type, since the pointer would be opaque. We need to account for the fact that a load of `last` in my example above must also load `arr[1]`, so optimizations that may want to split those up may need to be undone. These are the things the non-legacy cbuffer layout was trying to avoid, but we're stuck with what we have in DXIL for now.

In any case, we would still need the intrinsic to represent where the cbuffer comes from, so pretty well the only thing that would change here is that the `if (TypeName == "dx.CBuffer")` check would be some kind of pointer walk and address space check instead, so I'm not really sure that this PR is the best forum for discussing how we implement something that this doesn't yet implement anyway. I'm starting to work on a document describing what I think we should do about cbuffers in general that we will then be able to discuss this further on.